ledger/src/blockstore_db.rs

```diff
@@ -1207,6 +1214,28 @@ impl ColumnName for columns::Index {
 }
 impl TypedColumn for columns::Index {
     type Type = blockstore_meta::Index;
+
+    fn deserialize(data: &[u8]) -> Result<Self::Type> {
+        let config = bincode::DefaultOptions::new()
+            // `bincode::serialize` uses fixint encoding by default, so we need to use the same here
+            .with_fixint_encoding()
+            .reject_trailing_bytes();
+
+        // Migration strategy for new column format:
+        // 1. Release 1: Add ability to read new format as fallback, keep writing old format
+        // 2. Release 2: Switch to writing new format, keep reading old format as fallback
+        // 3. Release 3: Remove old format support once stable
+        // This allows safe downgrade to Release 1 since it can read both formats
+        // https://github.com/anza-xyz/agave/issues/3570
+        let index: bincode::Result<blockstore_meta::Index> = config.deserialize(data);
+        match index {
+            Ok(index) => Ok(index),
+            Err(_) => {
+                let index: blockstore_meta::IndexV2 = config.deserialize(data)?;
+                Ok(index.into())
+            }
+        }
```

Comment:
Building on top of what Zach and Alessandro mentioned, this is where there is "new behavior" in production.
- We read from Blockstore and try to deserialize into `Index` (the pre-existing type). If deserialize fails ...
    - Existing v2.1 and older code will panic if deserialize into `Index` fails (the caller unwraps this):
    https://github.com/anza-xyz/agave/blob/33ab693d94c787e64686e7b2d2f49492eae63885/ledger/src/blockstore_db.rs#L1730-L1733
    - This PR will instead try to deserialize into `IndexV2` ONLY if initial deserialize fails. But, seeing as we aren't writing the new format, that deserialize will fail too